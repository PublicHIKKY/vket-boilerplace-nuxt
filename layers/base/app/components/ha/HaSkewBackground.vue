<template>
  <div
    class="ha-skew"
    :style="backStyle"
  >
    <div
      class="content"
      :style="contentStyle"
    >
      <slot />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ValueOf } from '#base/app/utils/types/types'

const _AXIS = {
  x: 'x',
  y: 'y',
  z: 'z',
} as const

type Props = {
  deg: number
  axis: ValueOf<typeof _AXIS>
}

const props = defineProps<Props>()

const skew: Record<ValueOf<typeof _AXIS>, string> = {
  x: 'skewX',
  y: 'skewY',
  z: 'skewZ',
} as const

const backStyle = computed(() => ({
  transform: `${skew[props.axis]}(${props.deg}deg)`,
}))
const contentStyle = computed(() => ({
  transform: `${skew[props.axis]}(${props.deg * -1}deg)`,
}))
</script>

<style lang="scss" scoped>
@use '#base/app/assets/styles/variables' as v;
@use '#base/app/assets/styles/mixins' as m;

.ha-skew,
.content {
  height: 100%;
}
</style>
