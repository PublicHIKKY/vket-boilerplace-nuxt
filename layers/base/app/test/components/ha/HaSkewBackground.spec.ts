import { mount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'
import HaSkewBackground from '#base/app/components/ha/HaSkewBackground.vue'

describe('HaSkewBackground', () => {
  test('ref component', () => {
    expect(HaSkewBackground).toBeTruthy()
  })

  test('mount component', () => {
    const wrapper = mount(HaSkewBackground, {
      props: {
        deg: 30,
        axis: 'x',
      },
    })

    expect(wrapper.getCurrentComponent()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('slot', () => {
    const wrapper = mount(HaSkewBackground, {
      props: {
        deg: 30,
        axis: 'x',
      },
      slots: {
        default: 'Slot Content',
      },
    })

    expect(wrapper.text()).toContain('Slot Content')
  })

  describe('props', () => {
    test('renders props', () => {
      const wrapper = mount(HaSkewBackground, {
        props: {
          deg: 30,
          axis: 'x',
        },
      })

      expect(wrapper.exists()).toBe(true)
    })

    test('renders content with correct skew when axis is x', () => {
      const deg = 30
      const wrapper = mount(HaSkewBackground, {
        props: {
          deg,
          axis: 'x',
        },
      })

      expect(wrapper.find('.content').attributes('style')).toContain(
        `transform: skewX(${deg * -1}deg)`,
      )
    })

    test('renders content with correct skew when axis is y', () => {
      const deg = 45
      const wrapper = mount(HaSkewBackground, {
        props: {
          deg,
          axis: 'y',
        },
      })

      expect(wrapper.find('.content').attributes('style')).toContain(
        `transform: skewY(${deg * -1}deg)`,
      )
    })

    test('renders content with correct skew when axis is z', () => {
      const deg = 45
      const wrapper = mount(HaSkewBackground, {
        props: {
          deg,
          axis: 'z',
        },
      })

      expect(wrapper.find('.content').attributes('style')).toContain(
        `transform: skewZ(${deg * -1}deg)`,
      )
    })
  })
})
