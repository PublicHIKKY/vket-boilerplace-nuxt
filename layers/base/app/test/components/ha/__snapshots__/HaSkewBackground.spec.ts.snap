// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HaSkewBackground > mount component 1`] = `
"<div data-v-0570223f="" class="ha-skew" style="transform: skewX(30deg);">
  <div data-v-0570223f="" class="content" style="transform: skewX(-30deg);"></div>
</div>"
`;
